Delete app sessions on logout

gabrielcorado opened the pull request:
Closes #9124. It introduces a new gRPC endpoint, `DeleteUserAppSessions`, to delete all the user app sessions. This call is now being used on the logout flow. An integration test was added, it uses the reproduction steps described on the issue.

gabrielcorado committed:
feat: delete app web sessions during logout

rosstimothy commented on lib/auth/auth_with_roles.go:
```diff
@@ -3160,6 +3157,55 @@ func (a *ServerWithRoles) DeleteAllAppSessions(ctx context.Context) error {
+	for _, session := range sessions {
+		// The check is done against each session to avoid comparing user input
+		// with resources that the user doesn’t have access to.
+		if err := a.checkPermissionToDeleteUserWebSession(req.Username); err != nil {
```
This check is identical to the one on line 3163. How will this be any different inside this loop?

gabrielcorado replied:
It was wrong. The check here will be with the `session.GetUser()` instead of `req.Username`.

rosstimothy commented on lib/auth/auth_with_roles_test.go:
```diff
@@ -110,7 +110,6 @@ func TestGenerateUserCertsWithRoleRequest(t *testing.T) {
 	require.NoError(t, err)
 
 	dummyUserRole, err := types.NewRole("dummy-user-role", types.RoleSpecV4{})
-	require.NoError(t, err)
```
Was this accidental?

rosstimothy commented on lib/auth/auth_with_roles_test.go:
```diff
@@ -1620,3 +1619,102 @@ func TestNoElevatedAccessRequestDeletion(t *testing.T) {
+		require.NoError(t, err)
+		server, err := types.NewAppServerV3FromApp(app, "host", uuid.New().String())
+		require.NoError(t, err)
+		_, err = srv.Auth().UpsertApplicationServer(context.Background(), server)
```
```suggestion
		_, err = srv.Auth().UpsertApplicationServer(ctx, server)
```

rosstimothy commented on lib/services/local/session_test.go:
```diff
@@ -0,0 +1,66 @@
+)
+
+func TestDeleteUserAppSessions(t *testing.T) {
+	identity, clock := newIdentityService(t)
```
Create a local background context and replace all the `context.TODO` with it.

rosstimothy commented on lib/auth/auth_with_roles.go:
```diff
@@ -3160,6 +3157,55 @@ func (a *ServerWithRoles) DeleteAllAppSessions(ctx context.Context) error {
+		}
+
+		// A request will only delete web sessions from the `Username` provided.
+		if req.Username != session.GetUser() {
```
Should this check happen before `checkPermissionToDeleteUserWebSession`?

rosstimothy commented on lib/web/sessions.go:
```diff
@@ -698,6 +698,9 @@ func (s *sessionCache) invalidateSession(ctx *SessionContext) error {
 	if err != nil && !trace.IsNotFound(err) {
 		return trace.Wrap(err)
 	}
+	if err := clt.DeleteUserAppSessions(context.Background(), &proto.DeleteUserAppSessionsRequest{Username: ctx.user}); err != nil {
```
```suggestion
	if err := clt.DeleteUserAppSessions(context.TODO(), &proto.DeleteUserAppSessionsRequest{Username: ctx.user}); err != nil {
```

r0mant commented on lib/auth/auth_with_roles.go:
```diff
@@ -3160,6 +3157,55 @@ func (a *ServerWithRoles) DeleteAllAppSessions(ctx context.Context) error {
+
+// checkPermissionToDeleteUserWebSession checks if the current user can delete
+// WebSessions from the provided `username`.
+func (a *ServerWithRoles) checkPermissionToDeleteUserWebSession(username string) error {
```
Nit: A little too verbose I think, how about `canDeleteWebSession`?

r0mant commented on lib/auth/auth_with_roles.go:
```diff
@@ -3160,6 +3157,55 @@ func (a *ServerWithRoles) DeleteAllAppSessions(ctx context.Context) error {
+		return trace.Wrap(err)
+	}
+
+	sessions, err := a.authServer.GetAppSessions(ctx)
```
Can we either make a backend method that returns user's app sessions instead of checking `req.Username != session.GetUser()` below, or encapsulate it in the helper method? Something like `GetAppSessionsFor(ctx, <user>)`.

r0mant commented on lib/auth/auth_with_roles.go:
```diff
@@ -3160,6 +3157,55 @@ func (a *ServerWithRoles) DeleteAllAppSessions(ctx context.Context) error {
+	for _, session := range sessions {
+		// The check is done against each session to avoid comparing user input
+		// with resources that the user doesn’t have access to.
+		if err := a.checkPermissionToDeleteUserWebSession(session.GetUser()); err != nil {
```
Isn't this check redundant if we're only retrieving sessions for a particular user?

r0mant commented on lib/auth/auth_with_roles.go:
```diff
@@ -3160,6 +3157,55 @@ func (a *ServerWithRoles) DeleteAllAppSessions(ctx context.Context) error {
+func (a *ServerWithRoles) checkPermissionToDeleteUserWebSession(username string) error {
+	if err := a.currentUserAction(username); err != nil {
+		if err := a.action(apidefaults.Namespace, types.KindWebSession, types.VerbDelete); err != nil {
+			return err
```
```suggestion
			return trace.Wrap(err)
```

r0mant commented on lib/services/local/session.go:
```diff
@@ -89,6 +90,27 @@ func (s *IdentityService) DeleteAppSession(ctx context.Context, req types.Delete
 }
 
+// DeleteAllAppSessions removes all application web sessions.
+func (s *IdentityService) DeleteUserAppSessions(ctx context.Context, req *proto.DeleteUserAppSessionsRequest) error {
```
Is this Identity method actually used anywhere? In the ACL layer we still delete each session individually AFAICS. Is it needed to just satisfy some interface?

gabrielcorado replied:
I've changed the implementation to use this function on the ACL layer.

r0mant reviewed (commented):
@gabrielcorado With these changes, do we invalidate app sessions during password reset flow also?

smallinsky reviewed (commented):
Right now the user app session can be either created by web UI and tsh app login. 
Wonder if an app session generated by `tsh app login` should affected by Web UI Logout handler ?

gabrielcorado commented:
@rosstimothy @r0mant I've updated the logic to use `GetUserAppSessions` function internally, so the `DeleteUserAppSessions` doesn't need to filter sessions. Besides that, the `DeleteUserAppSessions` on the ACL layer now only checks user permission and calls the `DeleteUserAppSessions` from the backend.

@smallinsky As implemented, it will be affected (I added an integration test to validate that). Currently, I don't see any way of differentiating an AppSession created through Web or `tsh`. We could add a new field in the `WebSessionSpecV2` to identify where the session was created and pass this "origin" as a filter to `DeleteUserAppSessions`. What do you think?

r0mant commented on api/client/client.go:
```diff
@@ -1093,6 +1093,12 @@ func (c *Client) DeleteAllAppSessions(ctx context.Context) error {
 	return trail.FromGRPC(err)
 }
 
+// DeleteUserAppSessions delete all user’s application sessions.
```
```suggestion
// DeleteUserAppSessions deletes all user’s application sessions.
```

r0mant commented on api/client/proto/authservice.proto:
```diff
@@ -1563,6 +1567,8 @@ service AuthService {
     rpc DeleteAppSession(DeleteAppSessionRequest) returns (google.protobuf.Empty);
     // DeleteAllAppSessions removes all application web sessions.
     rpc DeleteAllAppSessions(google.protobuf.Empty) returns (google.protobuf.Empty);
+    // DeleteUserAppSessions delete all user’s application sessions.
```
```suggestion
    // DeleteUserAppSessions deletes all user’s application sessions.
```

r0mant commented on integration/app_integration_test.go:
```diff
@@ -1175,38 +1232,38 @@ func (p *pack) createAppSession(t *testing.T, publicAddr, clusterName string) st
+
+	req, err := http.NewRequest(method, u.String(), bytes.NewBuffer(payload))
+	if err != nil {
+		return 0, nil, err
```
```suggestion
		return 0, nil, trace.Wrap(err)
```

r0mant commented on integration/app_integration_test.go:
```diff
@@ -1175,38 +1232,38 @@ func (p *pack) createAppSession(t *testing.T, publicAddr, clusterName string) st
-
-	return casResp.CookieValue
+	statusCode, body, err := p.sendRequest(req, nil)
+	return statusCode, []byte(body), err
```
```suggestion
	return statusCode, []byte(body), trace.Wrap(err)
```

r0mant commented on lib/auth/auth_with_roles.go:
```diff
@@ -3193,6 +3190,32 @@ func (a *ServerWithRoles) DeleteAllAppSessions(ctx context.Context) error {
 	return nil
 }
 
+// DeleteUserAppSessions delete all user’s application sessions.
```
```suggestion
// DeleteUserAppSessions deletes all user’s application sessions.
```

r0mant commented on lib/auth/auth_with_roles.go:
```diff
@@ -3193,6 +3190,32 @@ func (a *ServerWithRoles) DeleteAllAppSessions(ctx context.Context) error {
+// WebSessions from the provided `username`.
+func (a *ServerWithRoles) canDeleteWebSession(username string) error {
+	if err := a.currentUserAction(username); err != nil {
+		if err := a.action(apidefaults.Namespace, types.KindWebSession, types.VerbDelete); err != nil {
```
In the case where we're deleting all sessions, this should also check for VerbList on web sessions.

r0mant commented on lib/services/identity.go:
```diff
@@ -266,6 +267,8 @@ type AppSession interface {
 	DeleteAppSession(context.Context, types.DeleteAppSessionRequest) error
 	// DeleteAllAppSessions removes all application web sessions.
 	DeleteAllAppSessions(context.Context) error
+	// DeleteUserAppSessions delete all user’s application sessions.
```
```suggestion
	// DeleteUserAppSessions deletes all user’s application sessions.
```

r0mant commented on lib/services/local/session.go:
```diff
@@ -89,6 +90,40 @@ func (s *IdentityService) DeleteAppSession(ctx context.Context, req types.Delete
+	return userSessions, nil
+}
+
+// DeleteAllAppSessions removes all application web sessions.
```
```suggestion
// DeleteAllAppSessions removes all application web sessions for a particular user.
```

r0mant commented on lib/web/sessions.go:
```diff
@@ -698,6 +698,9 @@ func (s *sessionCache) invalidateSession(ctx *SessionContext) error {
 	if err != nil && !trace.IsNotFound(err) {
 		return trace.Wrap(err)
 	}
+	if err := clt.DeleteUserAppSessions(context.TODO(), &proto.DeleteUserAppSessionsRequest{Username: ctx.user}); err != nil {
```
Should we also do this when user does `tsh app logout` in the CLI?

gabrielcorado replied:
Looking at the code, it does something similar by removing the application sessions one by one.

r0mant reviewed (commented):
Mostly looks good, just a few other comments/typos.

gabrielcorado committed:
Apply suggestions from code review

Co-authored-by: Roman Tkachenko <roman@goteleport.com>

gabrielcorado committed:
Merge branch 'master' into gabrielcorado/delete-websessions-on-logout

gabrielcorado committed:
refactor(auth): add VerbList action to delete user app sessions

r0mant reviewed (approved):
@gabrielcorado Sorry for delay here, for some reason I thought we already merged this. Let's get this in and backport to v9 pls.

gabrielcorado committed:
Merge branch 'master' into gabrielcorado/delete-websessions-on-logout

gabrielcorado committed:
tests(local): update function arguments

gabrielcorado committed:
Merge branch 'master' into gabrielcorado/delete-websessions-on-logout

gabrielcorado committed:
Merge branch 'master' into gabrielcorado/delete-websessions-on-logout

gabrielcorado committed:
Merge branch 'master' into gabrielcorado/delete-websessions-on-logout

gabrielcorado committed:
Merge branch 'master' into gabrielcorado/delete-websessions-on-logout

gabrielcorado committed:
Merge branch 'master' into gabrielcorado/delete-websessions-on-logout

gabrielcorado committed:
Merge branch 'master' into gabrielcorado/delete-websessions-on-logout

gabrielcorado committed:
Merge branch 'master' into gabrielcorado/delete-websessions-on-logout
